File: backend/app/schemas/generation.py
"""试卷生成模式."""

from datetime import datetime

from pydantic import BaseModel, Field


class GenerationCustomParams(BaseModel):
    """生成自定义参数."""

    difficulty_target: float | None = Field(default=None, ge=1.0, le=5.0)
    knowledge_focus: list[str] = []
    avoid_question_ids: list[int] = []
    prefer_real_questions: bool = True
    allow_ai_generation: bool = True
    count: int = Field(default=1, ge=1, le=10)
    title_override: str | None = None


class GenerationRequest(BaseModel):
    """生成请求."""

    template_id: int
    custom_params: GenerationCustomParams = GenerationCustomParams()


class GenerationTaskStatus(BaseModel):
    """生成任务状态."""

    task_id: str
    status: str  # 'pending'/'planning'/'retrieving'/'generating'/'validating'/'assembling'/'completed'/'failed'
    current_step: str | None = None
    progress_pct: float = 0.0
    paper_id: int | None = None
    error_message: str | None = None
    created_at: datetime | None = None


class GenerationWsMessage(BaseModel):
    """WebSocket 推送消息."""

    task_id: str
    status: str
    current_step: str | None = None
    progress_pct: float = 0.0
    detail: dict | None = None
    timestamp: datetime

File: frontend/src/types/generation.ts
/** 试卷生成类型定义 */

export interface GenerationCustomParams {
  difficulty_target?: number;
  knowledge_focus?: string[];
  avoid_question_ids?: number[];
  prefer_real_questions?: boolean;
  allow_ai_generation?: boolean;
  count?: number;
  title_override?: string;
}

export interface GenerationRequest {
  template_id: number;
  custom_params?: GenerationCustomParams;
}

export type GenerationStatus =
  | 'pending'
  | 'planning'
  | 'retrieving'
  | 'generating'
  | 'validating'
  | 'assembling'
  | 'completed'
  | 'failed';

export interface GenerationTaskStatus {
  task_id: string;
  status: GenerationStatus;
  current_step: string | null;
  progress_pct: number;
  paper_id: number | null;
  error_message: string | null;
  created_at: string | null;
}

export interface GenerationWsMessage {
  task_id: string;
  status: GenerationStatus;
  current_step: string | null;
  progress_pct: number;
  detail: Record<string, unknown> | null;
  timestamp: string;
}

export const GENERATION_STATUS_LABELS: Record<GenerationStatus, string> = {
  pending: '等待中',
  planning: '规划中',
  retrieving: '检索中',
  generating: '生成中',
  validating: '验证中',
  assembling: '组装中',
  completed: '已完成',
  failed: '失败',
};

export const AGENT_STEPS = [
  { key: 'planning', label: '规划', icon: '📋' },
  { key: 'retrieving', label: '检索', icon: '🔍' },
  { key: 'generating', label: '生成', icon: '✨' },
  { key: 'validating', label: '验证', icon: '✅' },
  { key: 'assembling', label: '组装', icon: '📦' },
] as const;
